src/parser.rs

```diff
@@ -1,58 +1,203 @@
+                    comment_node,
+                    source,
+                    file_type,
+                    file,
+                    is_inline,
+                ) {
+                    annotations.push(annotation);
                 }
-                None
-            } else {
-                None
             }
-        })
-        .collect())
+        }
+    }
+
+    Ok(annotations)
 }
 
+/// Parse annotation from line. Returns None if the syntax is wrong.
 fn parse_annotation(
     line: &str,
     line_number: usize,
+    comment_node: tree_sitter::Node,
+    source_code: &[u8],
+    file_type: &FileType,
     file: &Path,
     is_inline: bool,
 ) -> Option<Annotation> {
-    let at_pos = line.find('@')?;
+    let at_pos = line.find(TAG)?;
     let colon_pos = line[at_pos..].find(':')?;
 
     let kind = line[at_pos + 1..at_pos + colon_pos].trim().to_string();
     let content = line[at_pos + colon_pos + 1..].trim().to_string();
+    let context = extract_context(comment_node, source_code, file_type);
 
     Some(Annotation {
         kind,
         content,
+        context,
         location: Location {
             file: file.to_path_buf(),
             line: line_number,
             inline: is_inline,
         },
     })
 }
+
+/// Extract annotation context from tree sitter node.
+/// Traverses the tree from node to prent until it finds a significant (e.g. class definition,
+/// struct field, impl block.)
+fn extract_context(
+    comment_node: tree_sitter::Node,
+    source_code: &[u8],
+    file_type: &FileType,
+) -> CitationContext {
+    let mut current = comment_node;
+    let mut current_kind = current.kind().to_string();
+    let mut associated_name = None;
+    let mut variable_name = None;
+    let mut parent_type = String::new();
+
+    // Keep track of the most significant node found
+    let mut most_significant = None;
+
+    // Traverse up until we find a significant node
+    while let Some(parent) = current.parent() {
+        let parent_kind = parent.kind();
+
+        // Skip comment-related nodes
+        if parent_kind.contains("comment") {
+            current = parent;
+            continue;
+        }
+
+        // Update most significant node based on type
+        match parent_kind {
+            "class_definition" | "class_declaration" | "struct_item" => {
+                most_significant = Some((parent, parent_kind.to_string()));
+            }
+            "function_definition"
+            | "function_item"
+            | "function_declaration"
+            | "method_definition" => {
+                most_significant = Some((parent, parent_kind.to_string()));
+                break;
+            }
+            "assignment" | "let_declaration" | "variable_declaration" => {
+                if most_significant.is_none() {
+                    most_significant = Some((parent, parent_kind.to_string()));
+                }
+            }
+            "impl_item" => {
```

Comment:
This is great for now, but in future work it would be nice to separate the context-extraction logic out to a language-specific abstraction. Otherwise this part of the codebase could get very messy.

Let's wait and see if the syntax context is useful enough to justify the extra work.